### PR TITLE
Antoine/create inference image for listingclass

### PR DESCRIPTION
**Pull request by anotinelg**
**Description**

*JIRA ticket*: 

*What did you do?*

- forked from sagemaker-mxnet-inference-toolkit

- Add special docker directory for listing-class, where i created docker.cpu and docker.eia for the 1.4.1 and py3 framework/python.
- The docker.eia is a copy of the original one + some requirements needed for listingclass
- The docker.cpu is a copy of the original one + some changes in how we add dependency from sagemaker-inference-toolki (copied form the docker.eia files) + same requirements needed for listingclass. We have changed the way the sagemaker-inference-toolkit is imported because it fails with a version higher than 1.4.4
- We have added some script to build and push all the images.
- The tagging is done with the hash of the commit + possibility to add a postfix following semantic versioning.

*Is there any documentation?*

README

*How did you test your changes?*

- build and push all images

**Reviewer(s)** mouloud, Luca

**Deadline** 2020-06-23

**ETA** 20 minutes

**Expected Review**

*What do you want from the reviewers?*

- to understand the set of changes done on docker, look at the commit message https://github.com/letgoapp/sagemaker-mxnet-inference-toolkit/pull/1/commits/a91bda8e48eabdd40788ccef0f0aece94a4acd45  and https://github.com/letgoapp/sagemaker-mxnet-inference-toolkit/pull/1/commits/45b5ff0f6b3b18f4fb9223d907f0c880053b1dad
- validate decision on tagging (look the readme, and the scripts.  ./scripts/build_docker.py, scripts/push_docker.py). 
- test it (look Readme)

*Is there any specific part of the code you'd like to receive special attention?*

### CODE REVIEW

File: scripts/build_docker.py [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+    os.chdir(build_dir)
+
+    print('Building docker image in dir {}: {}'.format(build_dir, ' '.join(build_cmd)))
+    if not dry_run:
+        subprocess.check_call(build_cmd)
+
+    os.chdir(prev_dir)
+
+
+def get_git_version():
```
Comment (lcnletgo) [inline scripts/build_docker.py:64]: I'd call this function "get_git_commit" (when reading the name I thought "why do we need to know the git version?"

File: scripts/build_docker.py [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+            framework_version = args.eia_version if (arch == 'eia' and args.eia_version) else args.version
+            root_build_dir = os.path.join(args.docker_folder, framework_version)
+            tag = '{}-{}-py{}'.format(framework_version, tag_arch, py_version[0])
+
+            repo = '{}-eia'.format(args.repo) if arch == 'eia' else args.repo
+
+            tag_git = get_git_version()
+            semversion = "" if args.semversion is None else f"-v{args.semversion}"
+            prev = '{}:{}-latest'.format(repo, tag)
+            dest = '{}:{}-{}{}'.format(repo, tag, tag_git,semversion)
```
Comment (lcnletgo) [inline scripts/build_docker.py:96]: I thought we were using only one of tag_git or semversion (or use two tags, one with git commit and one with the semversion)

File: README.rst [hunk 3]
```diff
@@ -25,3 +25,44 @@ License
...
+- The AWS account parameter is used to extract already existing layers from previously pushed images.
+
+
+
+Example to build  all dockers existing in the docker-listingclass folder:
+```
+python scripts/build_docker.py --docker-folder docker-listingclass \
+   --account 491446337008 \
+   --repo preprod-ds-listingclass-inference \
+   --version 1.4.1
```
Comment (lcnletgo) [inline README.rst:50]: can we call "--version" "--mxnet-version"?

File: docker-listingclass/1.4.1/py3/Dockerfile.cpu [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+                                  sagemaker-inference==1.1.0 \
+                                  "sagemaker-mxnet-inference<2"
+
+
+# This is here to make our installed version of OpenCV work.
+# https://stackoverflow.com/questions/29274638/opencv-libdc1394-error-failed-to-initialize-libdc1394
+# TODO: Should we be installing OpenCV in our image like this? Is there another way we can fix this?
+RUN ln -s /dev/null /dev/raw1394
+
+ENV PYTHONDONTWRITEBYTECODE=1 \
```
Comment (lcnletgo) [inline docker-listingclass/1.4.1/py3/Dockerfile.cpu:57]: oh nice, we should get these environment variables in the EC2 instances where we do the training then :)
(for example for LANG and LC_ALL we use "en.UTF-8", I am not sure if that could cause use problems or not at some point.

File: docker-listingclass/1.4.1/py3/Dockerfile.cpu [hunk 5]
```diff
@@ -0,0 +1,88 @@
+FROM ubuntu:16.04
+
+LABEL maintainer="Amazon AI"
```
Comment (lcnletgo) [inline docker-listingclass/1.4.1/py3/Dockerfile.cpu:3]: Maybe we can change this to letgo? or datascience@letgo.com?

File: scripts/build_docker.py [hunk 6]
```diff
@@ -0,0 +1,105 @@
+# Copyright 2019-2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
```
Comment (lcnletgo) [inline scripts/build_docker.py:1]: @anotinelg just to understand, this is a script from AWS?

File: scripts/push_docker.py [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+# ANY KIND, either express or implied. See the License for the specific
+# language governing permissions and limitations under the License.
+from __future__ import absolute_import
+
+import argparse
+import subprocess
+import logging
+import os
+
+DEFAULT_REGION = 'us-east-1'
```
Comment (lcnletgo) [inline scripts/push_docker.py:20]: shouldn't we pass this as a parameter?

File: README.rst [hunk 8]
```diff
@@ -25,3 +25,44 @@ License
...
+   --semversion 0.1.0
+
+````
```
Comment (lcnletgo) [inline README.rst:53]: there's an extra "`" which doesn't make it render it correctly from Github

File: README.rst (hunk 3 above)
Comment (lcnletgo) [reply README.rst:50]: also it's missing a `\`

File: README.rst [hunk 9]
```diff
@@ -25,3 +25,44 @@ License
...
+
+
+
+Example to build  all dockers existing in the docker-listingclass folder:
+```
+python scripts/build_docker.py --docker-folder docker-listingclass \
+   --account 491446337008 \
+   --repo preprod-ds-listingclass-inference \
+   --version 1.4.1
+   --semversion 0.1.0
```
Comment (lcnletgo) [inline README.rst:51]: missing a `\`

File: README.rst (hunk 8 above)
Comment (lcnletgo) [reply README.rst:53]: I couldn't run the example, it complains about scripts/build_docker.py not existing, but the file does exist:

> ip-192-168-1-110:sagemaker-mxnet-inference-toolkit luca.cerone$ python scripts/build_docker.py --docker-folder docker-listingclass \
>    --account 491446337008 \
>    --repo preprod-ds-listingclass-inference \
>    --version 1.4.1 \
>    --semversion 0.1.0
  File "scripts/build_docker.py", line 43
    logging.warning(f"file {dockerfile_path} does not exist. Skipping")
                                                                     ^

> SyntaxError: invalid syntax
> ip-192-168-1-110:sagemaker-mxnet-inference-toolkit luca.cerone$ git branch
> \* antoine/create_inference_image_for_listingclass
>   master
> ip-192-168-1-110:sagemaker-mxnet-inference-toolkit luca.cerone$ 
> ip-192-168-1-110:sagemaker-mxnet-inference-toolkit luca.cerone$ ls scripts/build_docker.py 
> scripts/build_docker.py
> ip-192-168-1-110:sagemaker-mxnet-inference-toolkit luca.cerone$

File: README.rst [hunk 10]
```diff
@@ -25,3 +25,44 @@ License
...
+- The `semversion` may be empty.
+- the git hash is automatically extracted. If it exist any changes not commited, the git hash will be postfixed by
+a ``dirty` tag.
+- The AWS account parameter is used to extract already existing layers from previously pushed images.
+
+
+
+Example to build  all dockers existing in the docker-listingclass folder:
+```
+python scripts/build_docker.py --docker-folder docker-listingclass \
```
Comment (lcnletgo) [inline README.rst:47]: silly comment, on my Mac Python still defaults to Python 2.8 and the command was failing, maybe we can change `python`  to `python3` or mention explicitly in the Readme that the script works only with Python 3?

File: README.rst [hunk 11]
```diff
@@ -25,3 +25,44 @@ License
...
+The push_docker.py will push to AWS ECR all images defined in the docker-folder. It will not push images with the dirty
+ label.
+ For each image, it will also tag
+with the following name the image: `{repo}:{framework_version}-{arch}-{version python}-latest
+
+Push listingclass images:
+```
+python scripts/push_docker.py --docker-folder docker-listingclass \
+  --account 491446337008 --repo preprod-ds-listingclass-inference \
+   --version 1.4.1
```
Comment (lcnletgo) [inline README.rst:64]: `\` is missing

File: scripts/build_docker.py (hunk 1 above)
Comment (anotinelg) [reply scripts/build_docker.py:64]: ok

File: scripts/build_docker.py (hunk 2 above)
Comment (anotinelg) [reply scripts/build_docker.py:96]: i use both.. yes it is a little bit different to what we spoke one week ago

File: README.rst (hunk 3 above)
Comment (anotinelg) [reply README.rst:50]: ok

File: scripts/build_docker.py (hunk 6 above)
Comment (anotinelg) [reply scripts/build_docker.py:1]: It is a script that i have modified

File: scripts/push_docker.py (hunk 7 above)
Comment (anotinelg) [reply scripts/push_docker.py:20]: it is passed

**Comment by anotinelg** [comment]
> As discussed in person, maybe changing the error message,
> or clarifying a bit the expected behaviour would be good,
> but the code it is good to go (and for the versioning, we agreed that at the moment
> having the commit is more crucial than the version, and if needs change, we'll change the code in the future).

thanks @luca, i addressed your feedback in the following commit.
I will pushed the PR

**Pull request merged**